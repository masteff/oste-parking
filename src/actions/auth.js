import {auth} from '../firebase/firebase'

export const login = (uid, name) => ({
    type: 'LOGIN',
    uid,
    name
})

export const startLogin = (email, password) => {
    return () => {
        return auth.signInWithEmailAndPassword(email, password)
        .catch(e => console.log(e.message))
    }
}

export const logout = () => ({
    type: 'LOGOUT'
})

export const startLogout = () => {
    return () => {
        return auth.signOut()
        .catch(e => console.log(e.message))
    }
}

export const startCreate = (email, password, name) => {
    return (dispatch) => {
        auth.createUserWithEmailAndPassword(email, password)
        .then((res) => {
            res.user.updateProfile({
                displayName: name
            }).then(() => {
                dispatch(login(auth.currentUser.uid, auth.currentUser.displayName))
            })
        }).catch(e => console.log(e.message))
    }
}

export const startVerify = () => {
    return () => {
            const user = auth.currentUser
            if(user){
                return user.sendEmailVerification().then(() => {
                        console.log('email sent');
                }).catch(e => console.log(e.message))
       }
    }
}
