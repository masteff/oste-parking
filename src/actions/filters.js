import database from '../firebase/firebase'

export const setSpotOpen = (spotOpen) => ({
    type: 'SET_SPOT_OPEN',
    spotOpen
})

//setActualdate
export const setActualDate = (actualDate) => ({
    type: 'SET_ACTUAL_DATE',
    actualDate
})

export const getFreeDates = (freeDates) => ({
    type: 'GET_FREE_DATES',
    freeDates
})

export const startGetFreeDates = () => {
    return (dispatch) => {
        const filterAnswer = database.ref('freeDates').once('value').then((snapshot) => {
            dispatch(getFreeDates(Object.values(snapshot.val())))
        })
        return filterAnswer;

    }
}
