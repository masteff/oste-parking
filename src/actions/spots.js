import database from '../firebase/firebase'

//own spot
export const ownSpot = ({owner = '', id} = {}) => ({
    type: 'OWN_SPOT',
    id,
    owner
})

export const startOwnSpot = ({id} = {}) => {
    return (dispatch, getState) => {
        const owner = getState().auth.name
        const uid = getState().auth.uid
        dispatch(ownSpot({id,owner}))
        database.ref(`spots/${id}/owner`).set(owner)
        database.ref(`spots/${id}/ownerid`).set(uid)
    }
}
//give spot
export const giveSpot = ({id, freeOn} = {}) => ({
    type: 'GIVE_SPOT',
    id,
    freeOn
})

export const startGiveSpot = ({id, freeOn} = {}) => {
    return (dispatch) => {
        dispatch(giveSpot({id,freeOn}))
        return database.ref(`spots/${id}/freeOn`).push(freeOn).then(() => {
             database.ref('freeDates').push(freeOn).then(() => {
            })
        })
    }
}

export const getSpots = (spots) => ({
    type: 'GET_SPOTS',
    spots
})

export const startGetSpots = () => {
    return (dispatch) => {
        const spotAnswer = database.ref('spots').once('value').then((snapshot) => {
        const spots = []
            snapshot.forEach((child) => {
                spots.push({
                    id: child.key,
                    owner: child.val().owner,
                    ownerid: child.val().ownerid,
                    number: child.val().number,
                    freeOn: Object.values(child.val().freeOn), 
                })
            })
            dispatch(getSpots(spots))
        })
        return spotAnswer 
    }
}