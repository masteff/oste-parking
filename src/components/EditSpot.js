import React from 'react'

const EditSpot = (props) => (
    <div>
        Here I Will edit the spot with id {props.match.params.spot}
    </div>
)

export default EditSpot