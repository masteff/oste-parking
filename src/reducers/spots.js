import moment from 'moment'

//spotsReducer
const spotsDefaultState = []

export default (state = spotsDefaultState, action) => {
    switch (action.type){
        case 'GET_SPOTS':
            return action.spots
        case 'OWN_SPOT':
            return state.map((spot) => {
                if (spot.id === action.id){
                    return {
                        ...spot,
                        owner: action.owner
                    }
                } else {
                    return spot
                }
            })
        case 'GIVE_SPOT':
                return state.map((spot) => {
                    if (spot.id === action.id){
                        return {
                            ...spot,
                            freeOn: [...spot.freeOn, action.freeOn]
                        }
                    } else {
                        return spot
                    }
                })
        default:
            return state;
    }
}