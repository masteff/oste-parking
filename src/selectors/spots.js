// //Get free Spots
// export default (spots, {startDate, endDate}) => {
//     return spots.filter((spot) => {
//         const filterArray = spot.freeOn.filter((freeOn) => freeOn >= startDate && freeOn <= endDate)
//              return filterArray.length > 0
//     }).map((spot) => {
//          spot.freeOn.sort((a,b) => {
//              return a > b ? 1 : -1
//          })
//          return spot
//     }).sort((a,b) => {
//         return a.freeOn[0] > b.freeOn[0] ? 1 : -1
//     })
//  }