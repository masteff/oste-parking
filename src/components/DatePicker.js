import React, { useState } from 'react'
import { connect } from 'react-redux'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'
import { setActualDate } from '../actions/filters';

const DatePicker = (props) => {

    const [date, setDate] = useState(moment())

    const onDateChange = (date) => {
        if (date) {
            setDate(date)
            props.dispatch(setActualDate(date.valueOf()))
        }
    }

    const highlightDays = (day) => {
        return props.filters.freeDates.some((date) => day.isSame(moment(date), 'day'))
    }

    const onFocusChange = ({ focused }) => {

    }

    return (
        <SingleDatePicker
            date={date}
            onDateChange={onDateChange}
            focused={!props.filters.spotOpen}
            onFocusChange={onFocusChange}
            isDayHighlighted={highlightDays}
            keepOpenOnDateSelect={true}
            numberOfMonths={2}
            orientation='horizontal'
            anchorDirection='right'
            id='1'
        />
    )

}

const mapStateToProps = (state) => ({
    filters: state.filters
})

export default connect(mapStateToProps)(DatePicker)