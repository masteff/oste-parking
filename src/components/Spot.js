import React from 'react';
import { connect } from 'react-redux';
import {ownSpot, giveSpot, takeSpot} from '../actions/spots'
import getFreeSpots from '../selectors/spots'
import moment from 'moment'


export const Spot = ({number,owner,actualDate,freeOn,setOwner,setFree,setTaken,taken}) => (
    <div 
    style={freeOn.some((date) => date.isSame(actualDate, 'day')) ? 
    (taken.some((date) => date.takenOn.isSame(actualDate, 'day')) ? 
    divStyleTaken : divStyleFree) 
    : divStyle}
        >
        <p style={pStyle}>{number}</p>
        <p style={pStyle}>{owner}</p>
        <button onClick={() => {
                setOwner({owner:'Tiffany', number})
            }}>own</button>
        <button onClick={() => {
                setFree({number, freeOn: actualDate})
            }}>give</button>
        <button onClick={() => {
                setTaken({number, takenOn: actualDate, takenBy: 'Maddin'})
            }}>take</button> 
    </div>
)

const mapDispatchToProps = (dispatch) => {
    return {
        setOwner: (data) => dispatch(ownSpot(data)),
        setFree: (data) => dispatch(giveSpot(data)),
        setTaken: (data) => dispatch(takeSpot(data))
    }
}
const divStyle = {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    height: 150,
    width: 150,
    background: 'lightgrey',
    margin: 5
}

const divStyleFree = {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    height: 150,
    width: 150,
    background: 'lightgreen',
    margin: 5
}

const divStyleTaken = {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    height: 150,
    width: 150,
    background: 'red',
    margin: 5
}

const pStyle = {
    margin: 0
}


export default connect(undefined,mapDispatchToProps)(Spot)