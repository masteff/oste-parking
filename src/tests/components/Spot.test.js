import React from 'react'
import { shallow} from 'enzyme';
import {Spot} from '../../components/Spot'
import {ownSpot} from '../../actions/spots'
import {spots, filters} from '../fixtures'

const dispatch = jest.fn()


test('should render free Spot correctly', () => {
    const wrapper = shallow(<Spot {...spots[0]} {...filters} dispatch={dispatch}/>)
    expect(wrapper).toMatchSnapshot()
})

test('shoul set number and right', () => {
    const wrapper = shallow(<Spot {...spots[1]} {...filters} dispatch={dispatch}/>)
    const number = wrapper.find('p').at(0)
    expect(number.text()).toBe(spots[1].number.toString())
})

