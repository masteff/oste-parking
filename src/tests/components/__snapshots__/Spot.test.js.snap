// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render free Spot correctly 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "background": "lightgreen",
      "display": "flex",
      "flexDirection": "column",
      "height": 150,
      "justifyContent": "center",
      "margin": 5,
      "width": 150,
    }
  }
>
  <p
    style={
      Object {
        "margin": 0,
      }
    }
  >
    1
  </p>
  <p
    style={
      Object {
        "margin": 0,
      }
    }
  >
    Besitzer1
  </p>
  <button
    onClick={[Function]}
  >
    own
  </button>
  <button
    onClick={[Function]}
  >
    give
  </button>
  <button
    onClick={[Function]}
  >
    take
  </button>
</div>
`;
